File: 10-regular-expression-matching/regular-expression-matching.cpp
class Solution {
private:
    vector<vector<int>> memo;

    bool solve(string& s, string& p, int i, int j) {
        if (j == p.length())
            return i == s.length();
        if (memo[i][j] != -1)
            return memo[i][j];

        bool firstMatch = (i < s.length() && (s[i] == p[j] || p[j] == '.'));

        if (j + 1 < p.length() && p[j + 1] == '*') {
            memo[i][j] = (solve(s, p, i, j + 2) ||
                          (firstMatch && solve(s, p, i + 1, j)));
        } else {
            memo[i][j] = (firstMatch && solve(s, p, i + 1, j + 1));
        }
        return memo[i][j];
    }

public:
    bool isMatch(string s, string p) {
        memo.assign(s.size() + 1, vector<int>(p.size() + 1, -1));
        return solve(s, p, 0, 0);
    }
};

File: 10-regular-expression-matching/regular-expression-matching.go
func isMatch(s string, p string) bool {
    n, m := len(s), len(p)
    dp := make([][]bool, n+1)
    for i := range dp {
        dp[i] = make([]bool, m+1)
    }

    dp[n][m] = true

    for i := n; i >= 0; i-- {
        for j := m-1; j >= 0; j-- {
            var first_match bool = false
            if (i < n) && (s[i] == p[j] || p[j] == '.') { 
                first_match = true 
            }

            if (j + 1 < m) && p[j+1] == '*' {
                dp[i][j] = dp[i][j+2] || (first_match && dp[i+1][j])
            } else {
                dp[i][j] = first_match && dp[i+1][j+1]
            }
        }
    }
    return dp[0][0]
}